LC-9 connection request

bambooSocks opened the pull request:
I know it is not done yet ... two tests are not properly written, but I was wondering whether you could have a look at it now so I can see if I am doing the stuff correctly

bambooSocks committed:
Add basic model classes

bambooSocks committed:
Implement endpoint handling

bambooSocks committed:
Write test for ConnectionRequestService

Krismix1 commented on src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java:
```diff
@@ -0,0 +1,86 @@
+    @Value
+    public static class NewConnectionRequestDTO {
+        @NotNull
+        Long senderId;
```
We can/should also check for minimum value using something like `Min(1)`.

bambooSocks replied:
I have added the annotations

Krismix1 replied:
We could also write tests for negative values and for 0.

bambooSocks replied:
I already did

Krismix1 commented on src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java:
```diff
@@ -0,0 +1,86 @@
+        try {
+            connectionRequestService.delete(id);
+        } catch (NoSuchElementException e) {
+            return ResponseEntity.notFound().build();
```
I think typically the API should also return a message saying something like `Resource with id %d not found` (or maybe without the ID). This way the consumer of the API can differentiate between 404 meaning that the whole URL was not found or if just the resource with that ID.

Another way is to create a `ResourceNotFound` exception which will be raised in such scenarios. Then the global exception handler would catch it and return a 404 with the message. This could potentially keep the code more DRY.

bambooSocks replied:
I switched to `ResourceNotFoundException`

Krismix1 commented on src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java:
```diff
@@ -0,0 +1,86 @@
+        try {
+            connectionRequestService.accept(id);
+        } catch (NoSuchElementException e) {
+            return ResponseEntity.notFound().build();
```
The same here with the 404.

bambooSocks replied:
same here

Krismix1 commented on src/main/java/org/autopotato/linkedinclonebackend/repositories/CrudRepository.java:
```diff
@@ -0,0 +1,80 @@
+/*
```
Let's remove the license comment.

bambooSocks replied:
done

bambooSocks committed:
Finish ConnectionRequestService tests

bambooSocks committed:
Write tests for ConnectionController

bambooSocks committed:
Change code to fix PR

- Add `Min(1)` constrains to the NewConnectionRequestDTO fields
- Change No resource handling to use custom Exception
- Delete License from the CrudRepository

bambooSocks committed:
Add test for creating ConnectionRequest validation

Krismix1 commented on src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java:
```diff
@@ -0,0 +1,111 @@
+    public ResponseEntity<?> deleteConnectionRequest(@PathVariable long id) {
+        try {
+            connectionRequestService.delete(id);
+        } catch (ResourceNotFoundException e) {
```
Let's handle the exception in `ControllerExceptionHandler` as this will be a widely used exception.

bambooSocks replied:
ok I moved it

Krismix1 commented on src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java:
```diff
@@ -0,0 +1,111 @@
+        try {
+            connectionRequestService.delete(id);
+        } catch (ResourceNotFoundException e) {
+            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(e.getMessage());
```
The response body should contain a JSON object, not just a string.

bambooSocks replied:
I used the ExceptionResponseDTO

Krismix1 commented on src/test/java/org/autopotato/linkedinclonebackend/unit/controllers/ConnectionControllerTest.java:
```diff
@@ -0,0 +1,157 @@
+    @Test
+    final void createConnectionRequest() {
+        newConnectionRequestDTO =
+            new ConnectionController.NewConnectionRequestDTO(1L, 2L);
```
Do you need to explicitly say it's a long?

bambooSocks replied:
if you mean the literal then yes ... it complains that it is an integer since it takes `Long` not `long` ... no implicit conversion there from `int` :/

Krismix1 commented on src/test/java/org/autopotato/linkedinclonebackend/unit/services/ConnectionRequestServiceTest.java:
```diff
@@ -0,0 +1,110 @@
+
+    @BeforeEach
+    void setUp() {
+        MockitoAnnotations.initMocks(this);
```
What if we annotate the whole test class with `@ExtendWith(MockitoExtension.class)` instead of this?

bambooSocks replied:
I don't see a difference tho ... I was following tutorial and it worked ... but I guess for consistency I can change it

Krismix1 commented on src/main/java/org/autopotato/linkedinclonebackend/controllers/ResourceNotFoundException.java:
```diff
@@ -0,0 +1,15 @@
+
+@EqualsAndHashCode(callSuper = true)
+@Value
+public class ResourceNotFoundException extends Exception {
```
Usually, I would use the `RuntimeException` as the base class. But I can't say that there is a straightforward answer.
I guess we can try using caught exceptions and see how it goes.

For inspiration:
https://stackoverflow.com/questions/27578/when-to-choose-checked-and-unchecked-exceptions
https://stackoverflow.com/questions/499437/in-java-when-should-i-create-a-checked-exception-and-when-should-it-be-a-runti
https://stackoverflow.com/questions/6115896/understanding-checked-vs-unchecked-exceptions-in-java

bambooSocks replied:
I agree, I changed it to `RuntimeException` ... I always forget that it exists 🙄

bambooSocks committed:
Fix code by PR comments

Krismix1 commented on src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java:
```diff
@@ -0,0 +1,104 @@
+     */
+    @DeleteMapping("/requests/{id}")
+    public ResponseEntity<?> deleteConnectionRequest(@PathVariable long id)
+        throws ResourceNotFoundException {
```
We can probably remove the `throws ResourceNotFoundException` part(s) (also from tests). Is there a reason to keep them?

bambooSocks replied:
ok I removed it

bambooSocks committed:
Remove throws and switch for parametric tests

Krismix1 commented on src/test/java/org/autopotato/linkedinclonebackend/unit/controllers/ConnectionControllerTest.java:
```diff
@@ -0,0 +1,153 @@
+
+    private ObjectMapper jsonMapper;
+
+    public final NewConnectionRequestDTO[] conenctionOptions() {
```
There is a typo here in `conenctionOptions`. Also, I fail to find where this is used.

bambooSocks replied:
oh, I will handle that.. stand by

bambooSocks replied:
done

bambooSocks committed:
Remove redundant code

Krismix1 reviewed (approved):
Great work!
